tests/unit/lms/services/course_test.py

```diff
@@ -50,6 +51,73 @@ def test_get_by_context_id_with_no_match_and_raise_on_missing(self, svc):
         with pytest.raises(NoResultFound):
             svc.get_by_context_id("NO MATCH", raise_on_missing=True)
 
+    def test_get_from_launch_when_existing(
```

Comment:
These tests use `patch.object` which we tend to avoid. I reckon is the best tool here, we could instead:

- reproduce the data setup for get_by_context_id and upsert_course in each test
- create some sort of helper and assert calls to that. We'd make the service code worse for the benefit of this test.

Having service methods compose with each other seems like a good idea in any case.

Reply:
I think it's fine. If there is a nice split we could put in that's ideal, but if a service consumes it's own public methods I think it's fine to solve it like this.